File: C#/kmeans/kmeans/Cluster.cs
using System;
using System.Collections.Generic;
using System.Text;

namespace kmeans
{
    class Cluster
    {
        private Point centroid;
        private List<Point> points;

        public Cluster(Point centroid)
        {
            this.centroid = centroid;
            this.points = new List<Point>();
        }

        public void clearPoints()
        {
            this.points.Clear();
        }

        public void setPointList(List<Point> points)
        {
            this.points = points;
        }

        public int RecalculateCentroid()
        {
            Point newCentroid = getMean(this.points);
            if(diff(newCentroid, this.centroid) > 0)
            {
                this.centroid = newCentroid;
                return 0;
            }

            return 1;
        }

        public Point getMean(List<Point> points)
        {
            int i = 0;
            float[] mean = new float[points[0].getData().Length];
            foreach(Point point in points)
            {
                i = 0;
                foreach(float f in point.getData())
                {
                    mean[i] = mean[i] + f;
                    i++;
                }
            }
            i = 0;
            foreach (float f in mean)
            {
                mean[i] = MathF.Round( mean[i] / points.Count, 2);
                i++;
            }
            return new Point(mean);
        }
        private float diff(Point p1, Point p2)
        {
            float total = 0;
            float[] p2Data = p2.getData();
            int i = 0;
            foreach(float f in p1.getData())
            {
                total += Math.Abs(f - p2Data[i]);
                i++;
            }
            return total;
        }
    }
}

File: cpp/kmeans/kmeans/Cluster.cpp

#include <mutex>
#include "Point.cpp"
#include <cmath>
#include <iostream>
using namespace std;

class Cluster {
	Point centroid;
	vector<Point> points;
	static inline mutex mtx;

	float Diff(Point p1, Point p2, int len) {
		float total = 0;
		vector<float> p1Data = p1.GetData();
		vector<float> p2Data = p2.GetData();
		for (int i = 0; i < len; i++) {
			total += abs(p1Data[i] - p2Data[i]);
		}
		return total;
	}

public:

	Cluster() {

	}

	Cluster(Point centroid) {
		this->centroid = centroid;
		this->points = vector<Point>();
	}

	void AppendPoint(Point point) {
		mtx.lock();
		points.push_back(point);
		mtx.unlock();
	}

	void ClearPoints() {
		this->points.clear();
	}

	vector<Point> GetPoints() {
		return this->points;
	}

	int RecalculateCentroid() {
		int len = this->points.at(0).GetData().size();
		Point newCentroid = GetMean(this->points, len);
		if (Diff(newCentroid, this->centroid, len) > 0) {
			this->centroid = newCentroid;
			return 0;
		}
		return 1;
	}

	Point GetMean(vector<Point> points, int len) {
		vector<float> mean;
		for (int i = 0; i < len; i++) {
			mean.push_back(0);
		}

		for (int i = 0; i < points.size(); i++) {
			vector<float> pointData = points.at(i).GetData();
			for (int j = 0; j < len; j++) {
				mean[j] = mean[j] + pointData[j];
			}
		}
		for (int i = 0; i < len; i++) {
			mean[i] = round(mean[i] / points.size() * 100) / 100;	
		}
		Point nP(mean);
		return nP;
	}

	Point GetCentroid() {
		return this->centroid;
	}

};